unhandled dnspython error

/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/dns/name.py:334:_validate_labels(): A DNS label is > 63 octets long.



2024-06-24 11:47:08,149 [ERROR] bbot.scanner scanner.py:1079 Error in baddns.handle_event(URL_UNVERIFIED("https://www.hyundaimobil.co.id/news/details/beli-hyundai-ioniq-5-hemat-hingga-rp...", module=excavate, tags={'endpoint', 'spider-danger', 'in-scope'})): /home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/dns/name.py:334:_validate_labels(): A DNS label is > 63 octets long.
2024-06-24 11:47:08,151 [TRACE] bbot.scanner logger.py:83 Traceback (most recent call last):
  File "/home/liquid/bbot/bbot/scanner/scanner.py", line 1062, in _acatch
    yield
  File "/home/liquid/bbot/bbot/modules/base.py", line 629, in _worker
    await handle_event_task
  File "/home/liquid/bbot/bbot/modules/baddns.py", line 62, in handle_event
    if await task:
       ^^^^^^^^^^
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/baddns/modules/ns.py", line 37, in dispatch
    self.target_dnsmanager.answers["NS"] = await dnswalk.ns_trace(self.target)
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/baddns/lib/dnswalk.py", line 49, in ns_trace
    solved_nameservers = await self.ns_recursive_solve(nameserver_ips, target, depth=0)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/baddns/lib/dnswalk.py", line 88, in ns_recursive_solve
    query_msg = dns.message.make_query(target, dns.rdatatype.NS)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/dns/message.py", line 1787, in make_query
    qname = dns.name.from_text(qname, idna_codec=idna_codec)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/dns/name.py", line 1063, in from_text
    return Name(labels)
           ^^^^^^^^^^^^
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/dns/_immutable_ctx.py", line 40, in nf
    f(*args, **kwargs)
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/dns/_immutable_ctx.py", line 64, in __init__
    super().__init__(*args, **kwargs)
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/dns/name.py", line 373, in __init__
    _validate_labels(self.labels)
  File "/home/liquid/.cache/pypoetry/virtualenvs/bbot-N1wmp7PR-py3.11/lib/python3.11/site-packages/dns/name.py", line 334, in _validate_labels
    raise LabelTooLong
dns.name.LabelTooLong: A DNS label is > 63 octets long.

--- a/baddns/lib/dnswalk.py
+++ b/baddns/lib/dnswalk.py
@@ -47,6 +47,10 @@ class DnsWalk:
 
     async def ns_trace(self, target):
         log.debug(f"Attempting to find NS records for {target}")
+        for label in target.split("."):
+            if len(label) > 63:
+                log.debug(f"Skipping target [{target}]: label [{label}] exceeds 63-octet RFC 1035 limit")
+                return []
         nameserver_ips = self.root_servers[:]
         solved_nameservers = await self.ns_recursive_solve(nameserver_ips, target, depth=0)
         if solved_nameservers == None:

--- a/tests/ns_test.py
+++ b/tests/ns_test.py
@@ -65,3 +65,20 @@ async def test_ns_nosoa_generic(fs, configure_mock_resolver):
         "module": "NS",
     }
     assert any(expected == finding.to_dict() for finding in findings)
+
+
+@pytest.mark.asyncio
+async def test_ns_label_too_long(fs, configure_mock_resolver):
+    mock_data = {}
+    mock_resolver = configure_mock_resolver(mock_data)
+
+    target = "a" * 64 + ".bad.dns"
+    mock_signature_load(fs, "dnsreaper_wordpress_com_ns.yml")
+    signatures = load_signatures("/tmp/signatures")
+    baddns_ns = BadDNS_ns(target, signatures=signatures, dns_client=mock_resolver)
+
+    findings = None
+    if await baddns_ns.dispatch():
+        findings = baddns_ns.analyze()
+
+    assert not findings

